src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -1758,132 +1756,53 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserWithParams(
-                            if (isVisibleResponse) {
-                                responses.add(response);
-                                addEmailNamePairsToTable(emailNameTable, response,
-                                        question, roster);
-                                addEmailLastNamePairsToTable(emailLastNameTable, response,
-                                        question, roster);
-                                addEmailTeamNamePairsToTable(emailTeamNameTable,
-                                        response,
-                                        question, roster);
-                                addVisibilityToTable(visibilityTable, question,
-                                        response, userEmail, role, roster);
-                            }
-                        }
-                    }
-                }
-            }
 
-            addSectionTeamNamesToTable(sectionTeamNameTable, roster, courseId, userEmail, role,
-                                       feedbackSessionName, section);
-            
-            FeedbackSessionResultsBundle results =
-                    new FeedbackSessionResultsBundle(
-                            session, responses, relevantQuestions, emailNameTable,
-                            emailLastNameTable, emailTeamNameTable, sectionTeamNameTable,
-                            visibilityTable, responseStatus, roster, responseComments, true);
+        String section = params.get(PARAM_SECTION);
+        String questionId = params.get(PARAM_QUESTION_ID);
 
-            return results;
+        if (questionId != null) {
+            return getFeedbackSessionResultsForQuestionId(feedbackSessionName, courseId, userEmail, role, roster,
+                    session, relevantQuestions, allQuestions, isIncludeResponseStatus, section, questionId);
         }
 
+        // create empty data containers to store results
+        List<FeedbackResponseAttributes> responses = new ArrayList<FeedbackResponseAttributes>();
+
+        Map<String, String> emailNameTable = new HashMap<String, String>();
+        Map<String, String> emailLastNameTable = new HashMap<String, String>();
+        Map<String, String> emailTeamNameTable = new HashMap<String, String>();
+        Map<String, Set<String>> sectionTeamNameTable = new HashMap<String, Set<String>>();
+        Map<String, boolean[]> visibilityTable = new HashMap<String, boolean[]>();
+
+
         Map<String, FeedbackQuestionAttributes> allQuestionsMap = new HashMap<String, FeedbackQuestionAttributes>();
         for (FeedbackQuestionAttributes qn : allQuestions) {
             allQuestionsMap.put(qn.getId(), qn);
         }
         
         List<FeedbackResponseAttributes> allResponses = getAllResponses(feedbackSessionName, courseId,
-                params, section);
+                                                                        params, section);
         
         boolean isComplete = params.get(PARAM_RANGE) == null;
```

Comment:
It is also not a good thing that we declare it here, do all sorts of odd things with it, and then only use it 100 lines later, any idea how we can match the call site from the declaration site more closely?